### PR TITLE
feat: parse schema from SDL

### PR DESCRIPTION
**Pull request by JoviDeCroock**
This is an initial implementation of the [Schema Definition Language](https://spec.graphql.org/October2021/#sec-Type-System), currently this supports a schema like the following

```graphql
scalar MyScalar

type Todo {
  id: ID!
  text: MyScalar
}

type Query {
  todo: Todo!
}
```

It misses two SDL functionalities

- extending types
- directive definitions

This includes an initial set of validations that still need testing

### CODE REVIEW

File: src/schema/build_client_schema.rs [hunk 1]
```diff
@@ -237,16 +237,32 @@ pub(crate) mod private {
...
 
+            dbg!(self.name, &self.implementation.interfaces);
             if let Some(interfaces) = &self.implementation.interfaces {
                 for introspection_type_ref in interfaces.iter() {
                     let name = ctx.ctx.alloc_str(introspection_type_ref.name);
                     schema_interface_type.add_interface(ctx.ctx, name);
                 }
             }
 
+            dbg!(self.name, &self.possible_types.possible_types);
```
Comment (dpetrick) [inline src/schema/build_client_schema.rs]:
```suggestion
```

File: src/schema/build_client_schema.rs [hunk 2]
```diff
@@ -237,16 +237,32 @@ pub(crate) mod private {
                 schema_interface_type.add_field(ctx.ctx, schema_field);
             }
 
+            dbg!(self.name, &self.implementation.interfaces);
```
Comment (dpetrick) [inline src/schema/build_client_schema.rs]:
```suggestion
```

File: src/schema/introspection.rs [hunk 3]
```diff
@@ -154,7 +154,7 @@ pub enum IntrospectionInputTypeRef<'a> {
 #[cfg_attr(feature = "json", derive(Deserialize, Serialize))]
 #[cfg_attr(feature = "json", serde(rename_all = "camelCase"))]
 pub struct IntrospectionNamedTypeRef<'a> {
-    #[cfg_attr(feature = "json", serde(skip))]
+    #[cfg_attr(feature = "json", serde(borrow))]
```
Comment (dpetrick) [inline src/schema/introspection.rs:157]: Did you end up figuring out what's up here with the skips?

File: src/schema/sdl/finalizers.rs [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+            for obj in i.interfaces.iter() {
+                schema_interface.add_interface(ctx, obj);
+            }
+
+            for obj in i.possible_types.iter() {
+                schema_interface.add_possible_type(ctx, obj);
+            }
+
+            // TODO: possible interfaces aren't conveyed in the introspection
```
Comment (dpetrick) [inline src/schema/sdl/finalizers.rs:125]: Should there be post-processing to determine those? I' not sure what other SDL parsers do.

**Review by dpetrick — APPROVED** [review]

File: src/schema/sdl/finalizers.rs (hunk 4 above)
Comment (JoviDeCroock) [reply src/schema/sdl/finalizers.rs:125]: Possible interfaces actually doesn't exist in the reference implementation, not sure why we added that

**Pull request merged**